충남대 BE_김해경 2주차 (3단계)

HaegyeongKim01 opened the pull request:
위시 리스트를 만드는 과제였는데 역시 설계 부분에서 고민이 되었습니다. 
위시리스트를 사용자 Member객체가 List로 원하는 상품들을 가지고 있는 것이 맞을지 아니면 그냥 위시리스트 테이블 데이터베이스에서 쿼리로 위시리스트에 등록된 상품과 memberId를 조회하여 가져오는게 좋을지.. 이런 고민들이 되었으며 git 충돌 해결하거나 다루는 것이 미숙하여 지금부터 좀 더 git에 관련하여 공부하려 합니다!

HaegyeongKim01 committed:
feat: �¬�愿¢由 肄� 異�

HaegyeongKim01 committed:
feat: Service �대�� 異�

HaegyeongKim01 committed:
refactor: produc챙ct�� ProductDto濡 蹂¢寃

HaegyeongKim01 committed:
bug: service 二쇱� �ㅻ� �닿껐

HaegyeongKim01 committed:
feat: �¬�紐 寃¢利

HaegyeongKim01 committed:
validation �〈� 異�

HaegyeongKim01 committed:
[Step1] 湲곕 �뎄�ы 챙��¦

HaegyeongKim01 committed:
feat: 媛¢寃 �¦� �퓩≧� 寃¢�

HaegyeongKim01 committed:
Merge branch 'haegyeongkim01' into step1

HaegyeongKim01 committed:
style: BadRequestHandler 줄바꿈 추가

HaegyeongKim01 committed:
refactor: �¬�紐 �퓩≧� 寃¢� service 濡黝��¦ 泥━

HaegyeongKim01 committed:
[Step2] docs: 湲곕 �뎄�ы �¦

HaegyeongKim01 committed:
feat: �蕤��¦ 諛�� 媛庖껜 error 泥━

HaegyeongKim01 committed:
rename: index.html to admin.html

HaegyeongKim01 committed:
design: admin page

HaegyeongKim01 committed:
refactor: view 遺¦由

HaegyeongKim01 committed:
refactor: view 遺¦由

HaegyeongKim01 committed:
Merge branch 'step1' of https://github.com/HaegyeongKim01/spring-gift-wishlist into step2

HaegyeongKim01 committed:
feat: ���� �竈吏¢

HaegyeongKim01 committed:
feat: 濡帷렇� �竈吏¢

HaegyeongKim01 committed:
vifeat: ����, 濡帷렇� view 異�

HaegyeongKim01 committed:
feat: �� 媛庖껜, �ㅽ\ㅻ� 異�

HaegyeongKim01 committed:
feat: 濡帷렇� �몄�, �퓩¬ 諛�

HaegyeongKim01 committed:
feat: ���� 湲곕 援ы�

HaegyeongKim01 committed:
refactor: 濡帷렇� HTTP Status UNAUTHORIZED 蹂¢寃

HaegyeongKim01 committed:
Merge branch 'haegyeongkim01' into step2

HaegyeongKim01 committed:
docs: 援ы� �¦猷� 湲곕 list ��

HaegyeongKim01 committed:
refactor: 蹂¢寃�Controller to @RestController

HaegyeongKim01 committed:
refactor: Exception handler ��

HaegyeongKim01 committed:
refactor: Member default 沅� 異�

HaegyeongKim01 committed:
refactor: �퓩¬곗� 諛帷��쇱 異�

HaegyeongKim01 committed:
refactor: 蹂¢� static 蹂¢寃

HaegyeongKim01 committed:
refactor: 以¦諛� �帷굅

HaegyeongKim01 committed:
refactor: Member媛庖껜�¦ �대�� �몄�

�대�� 寃¢利 濡黝�� Member �대�� �대�濡 �대�� 愿¢� 濡黝�� � 怨녹�
紐⑥鯛�.
�ъ\ъK� 利�
媛¢�� 媛黝¦
�¦� �μ�� �린�¦濡 肄� 蹂¢寃

HaegyeongKim01 committed:
[Step3] docs: 湲곕 �뎄�ы �¦

HaegyeongKim01 committed:
[Step3] docs: 湲곕 �뎄�ы �¦

HaegyeongKim01 committed:
[Step3] docs: 湲곕 �뎄�ы �¦

HaegyeongKim01 committed:
Merge branch 'step3' of https://github.com/HaegyeongKim01/spring-gift-wishlist into step3

HaegyeongKim01 committed:
Merge branch 'step3' of https://github.com/HaegyeongKim01/spring-gift-wishlist into step3

HaegyeongKim01 committed:
rebase: step3 and step2

HaegyeongKim01 committed:
feat: �¦�蕤━�ㅽ ��� 湲곕

HaegyeongKim01 committed:
feat: �¦�蕤━�ㅽ ��� 湲곕

HaegyeongKim01 committed:
Merge branch 'step3' of https://github.com/HaegyeongKim01/spring-gift-wishlist into step3

HaegyeongKim01 committed:
refactor: role �� NOT NULL 異�

HaegyeongKim01 committed:
refactor: role �� NOT NULL 異�

HaegyeongKim01 committed:
refactor: �� message ��

HaegyeongKim01 committed:
refactor: �� message ��

HaegyeongKim01 committed:
feat: Constructor 추가

HaegyeongKim01 committed:
Merge branch 'haegyeongkim01' into step3

HaegyeongKim01 commented:
conflicts가 발생하여 step3에서 step2로 rebase 시도 했었는데 이 부분에서 좀 꼬여서 일단 write access로 충돌 제거 하였습니다. rebase 방법을 더 공부해보도록 하겠습니다!

HaegyeongKim01 committed:
Merge branch 'step3' of https://github.com/HaegyeongKim01/spring-gift-wishlist into step3

Hyeon9mak commented on README.md:
```diff
@@ -1,5 +1,20 @@
+### 기능 요구 사항
+- [ ] 위시 리스트에 등록된 상품 목록 조회
+- [ ] 위시 리스트에 상품 추가
+- [ ] 위시 리스트에 상품 제거
```
👀

Hyeon9mak commented on src/main/java/gift/controller/WishlistController.java:
```diff
@@ -0,0 +1,67 @@
+        return service.getWishProductLost(memberId);
+    }
+
+    @GetMapping("/wishlist/add")
+    public ResponseEntity<Void> addToWishlist(@RequestParam("productId") Long productId, @RequestHeader("Authorization") String authorizationHeader) {
+        String token = getBearerToken(authorizationHeader);
+        String memberId = jwtUtil.getMemberEmailFromToken(token);
+
+        WishProduct wishProduct = new WishProduct(memberId, productId);
+        Boolean result = service.addWishProduct(wishProduct);
+        if (result) {
+            return ResponseEntity.ok().build();
+        }
+        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
+    }
```
HTTP 요청에서 method 는 어떤 역할을 갖는다고 생각하시나요?! 🤔 
https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET

HaegyeongKim01 replied:
아! 실수했습니다! 
위시리스트 조회 - GET
추가 - POST
삭제 - DELETE

Hyeon9mak commented on src/main/java/gift/controller/WishlistController.java:
```diff
@@ -0,0 +1,67 @@
+        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
+    }
+
+    /**
+     *
+     * @param authorizationHeader Authorization 헤더
+     * @return Bearer 토큰 추출
+     */
+    private static String getBearerToken(String authorizationHeader) {
+        return authorizationHeader.replace("Bearer ", "");
+    }
```
토큰 추출은 jwt util 을 다루는 다른 객체에게 명령을 보내볼 수 있지 않을까요? 🤔

HaegyeongKim01 replied:
넵! JwtUtil에서 메소드를 가지도록 변경하였습니다.🙂

Hyeon9mak commented on src/main/java/gift/vo/WishProduct.java:
```diff
@@ -0,0 +1,30 @@
+
+public class WishProduct {
+
+    private Long wishProductId;
+    private String memberId; //member pk is email
+    private Long productId;
```
필드명은 memberId 인데, pk 가 email 이라는게 어떤 의미인지 잘 이해하지 못했어요!

HaegyeongKim01 replied:
WishProduct의 외래 키(FK)로 Member의 기본 키(PK)를 사용하게 되는데, 이때 memberId가 단순한 숫자(예: 정수, Long)가 아닌 이메일로 사용되고 있음을 설명하고 싶습니다. 필드명을 포괄적으로 하여 데이터베이스 구조가 변경되었을 때 코드의 일관성을 유지하기 위함으로 memberId라고 작성했습니다. 하지만 Member PK가 바뀔 수 있습니다. 만약 Member의 PK 타입이 변경된다면, WishProduct의 해당 필드 타입도 그에 맞춰 변경되어야 하는데 이를 설명하고 싶었습니다.

Hyeon9mak commented on src/main/resources/db/h2/schema.sql:
```diff
@@ -14,5 +14,14 @@ CREATE TABLE PRODUCT (
+-- Create the WishProduct
+CREATE TABLE WISH_PRODUCT (
+                          wish_product_id BIGINT AUTO_INCREMENT PRIMARY KEY,
+                          member_id VARCHAR(100),
+                          product_id BIGINT,
```
member id 와 product id 도 not null 하게 다룰 수 있지 않을까요?

Hyeon9mak commented on src/main/resources/db/h2/schema.sql:
```diff
@@ -14,5 +14,14 @@ CREATE TABLE PRODUCT (
+                          wish_product_id BIGINT AUTO_INCREMENT PRIMARY KEY,
+                          member_id VARCHAR(100),
+                          product_id BIGINT,
+                          CONSTRAINT fk_member FOREIGN KEY (member_id) REFERENCES MEMBER(email),
+                          CONSTRAINT fk_product FOREIGN KEY (product_id) REFERENCES PRODUCT(id)
```
fk 를 만드는 것과 만들지 않는 것에는 어떤 장단점이 있을까요?

HaegyeongKim01 replied:
fk를 만들면 객체끼리의 연관관계가 생기기 때문에 데이터 무결성이라는 장점과 테이블을 분리하고 FK 하나로 연결 시킬 수 있기에 분리 가능한 장점이 있다고 생각합니다. 그런데 단점은 테이블 구조를 변경하거나 제약이 걸리는 경우가 생기는 단점이 존재합니다. 
추가적으로 찾아보니 테이블 구조 변경, 테스트 데이터를 만드는 것이 불편하다 등의 단점이 있지만 외래키 컬럼에 인텍스를 달아주고 부모 테이블의 PK가 변하지 않도록 디자인 하는 것으로 외래키를 사용하는 것이 더 나으며 성능적인 문제는 없다고 생각할 수 있다고 합니다.

Hyeon9mak replied:
여러가지 장단점이 있겠지만 대표적으로 제가 체감한 것은...

#### 장점
- RDB 라는 특성상 데이터 정합성과 관계표현이 가장 중요한데 이것을 잘 드러내준다.
- 테이블을 처음 확인하는 동료 개발자도 테이블간 관계를 바로 파악할 수 있다.
- 데이터가 입력/수정/삭제될 때 정합성을 보장할 수 있다.

#### 단점
- 데이터가 입력/수정/삭제될 때마다 정합성을 검사하므로 다루는 데이터 규모가 커지면 성능상 문제가 생길 수 있다.
- 서비스를 운영하다보면 데이터를 수기로 관리하는 일이 분명 생기는데, 이 때 FK 가 큰 걸림돌이 된다.

와 같은 장단점이 있었습니다.
현재는 규모가 작고, 먼 훗날 자연스럽게 FK 제거를 경험하게 되실테니 지금처럼 FK 를 적극 활용해보시는게 좋겠네요!

Hyeon9mak reviewed (changes requested):
안녕하세요 해경님, 3단계 잘 진행해주셨어요.
간단한 코멘트 남겼으니 확인해주세요!

HaegyeongKim01 committed:
docs: �뎄�ы 泥댄¬щ�� ��

HaegyeongKim01 committed:
refactor: JwtUtil �퓩¬ 異� 硫¦蕤� 蹂¢寃

WishController�¦ JwtUtil � �퓩¬ 異� 硫�蕤� JwtUtil �대��ㅼ��
媛¢吏¢寃 �⑥矧濡 JwtUtil 怨 愿¢�⑤� 硫� ���¦ �μ‾¬, �ъ\ъK 利�� �μ�
議댁頡

HaegyeongKim01 committed:
style: 二쇱¦ 蹂댁땐 �ㅻ�

HaegyeongKim01 committed:
refactor: wishlist DB �ㅽ\ㅻ� 蹂¢寃

HaegyeongKim01 committed:
refactor: HTTP method 蹂¢寃

HaegyeongKim01 committed:
refactor: HTTP 硫¦蕤� 蹂¢寃

HaegyeongKim01 committed:
refactor: URL Mapping 寃쎈� 蹂¢寃

HaegyeongKim01 committed:
style: 二쇱¦ 蹂댁땐 �ㅻ�

Hyeon9mak commented on src/main/java/gift/vo/WishProduct.java:
```diff
@@ -0,0 +1,30 @@
+
+public class WishProduct {
+
+    private Long wishProductId;
+    private String memberId; //Member pk는 String-email. Member 데이터 구조가 변경된다면 이 필드 타입도 변경 필요
+    private Long productId;
```
의도를 이해했습니다!
그렇다면 우선 처음보는 사람도 어떤 형태의 필드인지 이해할 수 있도록 `email` 이라는 명칭을 사용하고, 주석을 통해 pk 임을 드러내 주는건 어떨까요? 😉

Hyeon9mak reviewed (approved):
안녕하세요 해경님! 피드백 잘 반영해주셨습니다 👍 
2주차도 고생 많으셨습니다. 간단한 코멘트 남겼으니 확인해주세요.
3주차 미션에서 만나요 😄
